---
title: "SonicHedgehog Vignette"
output: rmarkdown::html_vignette
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry{SonicHedgehog Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


*Collaborators: Savvy Cornett & Jyhreh Johnson*

# Background
Given all life on earth diverged from a common ancestor, we need a way to analyse those evolutionary relationships. We can do this by comparing the sequences--nucleotide for DNA and RNA, or amino acid (polypeptide) for protein--of genes from different species. This can be done with whole genomes or otherwise highly conserved genes in the genome to determine the evolutionary relationship of the species, or with just a single gene/protein from multiple species to determine the relationship and amount of conservation for just that gene/protein. These comparisons can be visualized as a phylogenetic tree, which uses relative branch lengths to indicate time since species diverged from their common ancestor based on the number of substitutions (silent/missense/nonsense mutations) within the sequence being compared. Phylogenetic Comparative Analysis or Phylogenetic Comparative Methods (PCMs) are a set of statistical methods used by biologist, geneticist, and anthropologists to analyze phylogenetic trees and track/present species traits (Pennell & Harmon, 2013). Phylogenetic trees allow us to estimate the time for various evolutionary events, and facilitate the analysis of evolutionary processes, such as selection, gene flow, and speciation.\

# Introduction\
In this package, we have included a dataset for the Sonic Hedgehog vertebrate orthologs (the dataset: `sonic`) from the NIH database (NIH), as well as the peptide sequences in that dataset as a fasta file (`Shh_fasta` and `Shh_fasta.f`). We also have 5 functions:\
1) `read_file()`: reading in a .csv from a character vector\
2) `AlignAA()`: aligns multiple peptide sequences in fasta formatted data\
3) `phyloAA()`: calculates the pairwise distance between AA and clusters species/sequences with the neighbor joining algorithm\
4) `treeAA()`: plots a phylogenetic tree, based on {ggtree}\
5) `plot_Tree()`: plots maximum likelihood phylogenetic tree\

```{r setup}
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("ggtree")
BiocManager::install("msa")

library(sonichedgehog)
```


## Package Datasets\
A.`sonic`\
For this project, we have included an example dataset to use with our PCMs package, **sonic**, sonic is a dataset of Sonic hedgehog signalling molecule (Shh) orthologs from species of order Carnivora, a group of placental mammals specialized in eating flesh. The outgroup of this dataset is the insectivore *Erinaceus europaeus* (Western European hedgehog). These orthologs were retrieved from NCBI gene database. This gene plays a very important role in body patterning of the ventral neural tube and anterior-posterior limb axes during embryogenesis in vertebrates and invertebrates. Included in it are the following for each species, 25 total: GeneID, GeneSymbol, Scientific_name, Common_name, Description of the polymer molecule, RefSeq_Transcript_Accessions, Base_pairs (count) in gene,  RefSeq_Protein_Accessions, Amino_acids (count) in protein, and Peptide_Sequences from the respective protein accession.\


```{r}
sonic #prints the included sonic dataset
#library(usethis)
#ldata ready to load also included in the package
#sonic.f # sonic the raw file git link (.f)
```

As an alternative to the built-in `sonic` dataset, you can also use a function to read in the data. The function, read_file(), used in the following section of code reads in the raw/.csv file data necessary for the package. This function makes it easier for the user to load in data and cleans up the code. The function uses the f <- "file_link" & d <- read_csv() code.
```{r}
#use the function to read in the data and create the r dataset
sonic.f <- "https://raw.githubusercontent.com/slcornett/sonichedgehog/master/data/sonic.csv"
sonic.csv <- read_file(sonic.f)
head(sonic.csv) # same as sonic
```

B. `Shh_fasta`\
Additionally, the amino acid (peptide) sequences for each species Shh singaling molecule is included in fasta format in the file **Shh_fasta**. This sequences are the same as those included in `sonic`, but in the .fasta format. This file is a more accessible way to use the peptide sequences from the sonic dataset as an example.\
```{r Shh_fasta}
Shh_fasta # shows the dataset inclused in the package
class(Shh_fasta) # "AAStringSet"


Shh_fasta.f #raw file git link (.f)
```

# Phylogenetic analysis\
## **By Hand**\
For this example, we will be using the sonic fasta file included in this package (**Shh_fasta**).\
**1. convert the .fasta file data to a string of amino acids (AA), i.e., a literal polypeptide chain in code.**\
To do this, we first assign the link to the raw file to a value, then, using the `readAAStringSet()` function from the {Biostrings} package, convert the fasta file to an amino acid string set for each species included in the file (25 total). You can choose to keep the fasta format for the data with the `format = "fasta"` argument.\
```{r step 1 by hand}
#for reading multiple AA sequences from msa package
Shh_fast.f <- readAAStringSet(Shh_fasta.f, format = "fasta", use.names = TRUE) # format: biostrings, AAString set
Shh_fast.f # shows raw file details: for each, shows length of the sequence, the sequence, and seq name
```
We can confirm our file was correctly converted using the base ***R*** `class()` function. The class should be "AAStringSet", attributed to the "Biostrings" package.\
```{r step 1 check class}
# EXPECTED OUTPUT:
# "AAStringSet"
# attr(,"package")
# "Biostrings"
class(Shh_fast.f)
```

**2. Multiple Sequence Alignment (MSA)**\
We next need to conduct a multiple sequence alignment (msa)--the alignment of 3 or more sequences of genetic material (DNA, RNA, or protein) to determine homology and evolutionary relationships--on our Shh amino acid sequences (from `Shh_fasta` built into the package). We can do this using the `msa()` function from the {msa} package (package available from BioConductor; Bodenhofer et al., 2015). This function gives us the option to use one of three different algorithms, depending on the type of sequences being aligned. We will use the *MUSCLE* algorithm for our Shh protein sequences (from `Shh_fasta`) as this algorithm is especially good with proteins and our sequences are a medium/small size. Other algorithms options in the `msa()` function include *ClustalW* and *ClustalOmega*, both of which are good for all types of genetic material, especially those with medium to long sequences and many sequences being compared. You can learn more about multiple sequence alignments at the EMBL European Bioinformatics Institute's website: <https://www.ebi.ac.uk/Tools/msa/>.\
In this function, we define the method as `method = c("Muscle")` since we are using the *MUSCLE* algorithm, and the type of sequence as `type = "protein`. The order argument refers to the order of the sequences in the output object. Here, the sequences will be ordered by how aligned they are then the input order (ie their order in the fasta file): `order = c("aligned", "input")`.\
```{r step 2 by hand msa}
library(msa)
# multiple sequence alignment from msa package 
#align the fasta file using MUSCLE algorithm
fas_msa <-  msa(Shh_fasta, method = c("Muscle"), cluster = "default", type = "protein", verbose = TRUE) 
# print the complete results, with the consensus sequence
print(fas_msa, showConsensus = TRUE, show = "complete") 
```
>NOTE: if you run the `msa()` function more than 16 times in a single session (defined in the function as maxiter, for maxiumum iterations), will stop working (*SetIdCount Error*). To fix this, save everything, close the r-session, then re-open the r-session. It will work again after this.\

**2A.** check the class of the msa output:\
```{r check msa class}
# EXPECTED OUTPUT:
# "MsaAAMultipleAlignment"
# attr(,"package")
# "msa"
class(fas_msa)
#we no longer need the msa package so can detach it
```


The rest of the by-hand steps will use functions from the package {ape}: {ape}, Analyses of Phylogenetics and Evolution, is a really great package for writing, manipulating DNA and Amino Acid sequences for phylogenetic analysis (Paradis & Schliep, 2019).\

**3. Storing data in an bin**\
Here we will be storing the now `MsaAAMultipleAlignment` data as a matrix in an AAbin (amino acid bin) using the function `as.AAbin()`from the package {ape}. This function changes the class of the multiple sequence alignment data to class `AAbin`.\
The argument `show.aa =` will show the AA symbols or not (FALSE by default), and `check.names =` is a logical to check the row names before binding the sequences to the species names (the default is to only keep sequences/names with no missing data, and those with missing data will be dropped.) Since our data has already been aligned, they are all the same length, so we want `check.names = TRUE` (Paradis & Schliep, 2019).\
```{r Step 3 AAbin}
library(ape)
fas_AAbin <- as.AAbin(fas_msa, show.aa = TRUE, check.names = TRUE)
fas_AAbin # prints the details of the info stored in the AAbin matrix
class(fas_AAbin) # AAbin
```


**4. Calculating the phylogenetic distances between sequences**\
In order to to make a phylogenetic tree with our data, we need to calculate the number of AA differences between each pair of sequences in a matrix. These distances will inform the branch lengths of our phylogenetic tree. **NOTE:** This function can only be run on data in an "AAbin". Within this function are two arguments:\
- ` pairwise.deletion =` a logical indicating whether to delete the sites with missing data in a pairwise way (by default, sites with at least one missing data for all sequences are deleted), and\
- `scaled =` a logical value specifying whether to scale the number of AA differences by the sequence length (Paradis & Schliep, 2019).\
```{r AA distances}
dist_fas_AAbin <- dist.aa(fas_AAbin, pairwise.deletion = TRUE, scaled = TRUE)
dist_fas_AAbin #prints the scaled distances matrix
class(dist_fas_AAbin) #dist matrix
```


**5. Make a phylogenetic tree using the Neighbor-Joining Method**\
The Neighbor-Joining algorithm approximates the least squares trees (minimum evolution), and clusters nodes similar to each other and dissimilar from other clusters, based on the distances calculated in the last step (Saitou & Nei, 1987).\
```{r nj tree, fig.height=8, fig.width=8}
tree <- nj(dist_fas_AAbin)
tree # details about the phylogenetic tree (unrooted)
class(tree) #phylo 

# Add outgroup at node as Erinaceus_europaeus_Shh_OUTGROUP
# by first viewing the tree$tip.label for all 26 tips in the tree, we can see which node the outgroup is on
# make a vecotr of the tip.label that is the outgroup
outgroup <- tree$tip.label[2]
# root the tree on the outgroup
tree <- root(tree, outgroup = outgroup, node = NULL, resolve.root = TRUE)
is.rooted(tree) #TRUE
tree # updated details about the phylogenetic tree (rooted on the OUTGROUP)
```


**6. Plotting Tree with ggtree**\
You have many options when it comes to plotting a phylogenetic (class = `phylo`) tree. You may opt for the very simple base `plot()` function...\
```{r plot tree, fig.height=8, fig.width=8}
library(ggtree)
library(ggplot2)
# base R plot
plot(tree)
```
... or you can customize a tree to be green using {ggtree}\
```{r ggplot tree, fig.height=8, fig.width=8}
# ggtree, an extension package of ggplot2
ggt <- ggtree(tree, cex = 1, aes(color = branch.length), layout = "roundrect") +
  scale_color_continuous(high = 'green', low = 'darkgreen') + 
  #scale_color_continuous(type = "viridis") +
  theme(legend.position = "left") +
  geom_tiplab(align = FALSE, size = 3) +
  geom_treescale(x = 0, y = 0, color = "black", fontsize =3)
ggt
```


## **Using {sonichedgehog} functions**\
So many steps, right? Wouldn't it be nice to do less steps and arrive at the same endpoint? Now you can!\
### Align the Peptide Sequences\
Here we are using the `Shh_fasta.f`, file (class = "character") version of `Shh_fasta` included in this package to run it through the {sonichedgehog} function `AlignAA()`. This function reads in converts the raw fasta file to an AAstring, does a multiple sequence alignment using the MUSCLE algorithm on the fasta data, then prints the complete set of aligned AA seqeuences with the species names and consensus sequence. Essentially, this combines by-hand steps 1-2. 
```{r AlignAA}
Shh_fasta.f # character vector containing the git link
# read in the vector the raw git link or dataset is assigned to
pat <- AlignAA(Shh_fasta.f)
pat # prints the MUSCLE alignment 
class(pat) # "MsaAAMultipleAlignment"
```

### Running and Plotting PCMs\
**Distance and NJ with `phyloAA()` function**
```{r phyloAA}
ptree <- phyloAA(pat)
ptree 
class(ptree)

# Add outgroup at node as Erinaceus_europaeus_Shh_OUTGROUP
# by first viewing the tree$tip.label for all 26 tips in the tree, we can see which node the outgroup is on
# make a vecotr of the tip.label that is the outgroup
outgroup <- ptree$tip.label[2]
# root the tree on the outgroup
ptree <- root(ptree, outgroup = outgroup, node = NULL, resolve.root = TRUE)
is.rooted(ptree) #TRUE
ptree # updated details about the phylogenetic tree (rooted on the OUTGROUP)
```

**Plotting with `treeAA()` function**
```{r treeAA, fig.height=8, fig.width=8}
tree_plot <- treeAA(ptree)
tree_plot
```

**Plotting Maximum Likelihood with `plotTree()` function**
```{r plot_Tree, fig.height=8, fig.width=8}
pt <- plot_Tree(tree)
```

## References\ 
Bodenhofer, U., Bonatesta, E., Horejš-Kainrath, C., & Hochreiter, S. (2015). Msa: An r package for multiple sequence alignment. *Bioinformatics*, btv494. <https://doi.org/10.1093/bioinformatics/btv494>.\
- Edgar, R. C. (2004) MUSCLE: multiple sequence alignment with high accuracy and high throughput. *Nucleic Acids Res. 32*(5):1792-1797. DOI: 10.1093/nar/gkh340.\
- Edgar, R. C. (2004) MUSCLE: a multiple sequence alignment method with reduced time and space complexity. *BMC Bioinformatics 5*:113. DOI: 10.1186/1471-2105-5-113.\

Pennell, M. W., & Harmon, L. J. (2013). An integrative view of phylogenetic comparative methods: Connections to population genetics, community ecology, and paleobiology: Integrative comparative methods. *Annals of the New York Academy of Sciences, 1289*(1), 90–105. <https://doi.org/10.1111/nyas.12157>.\

NCBI. SHH orthologs. (2022). [Database]. NCBI; National Library of Medicine. <https://www.ncbi.nlm.nih.gov/gene/6469/ortholog/>.\

Paradis, E., & Schliep, K. (2019). Ape 5.0: An environment for modern phylogenetics and evolutionary analyses in r. *Bioinformatics, 35*(3), 526–528. <https://doi.org/10.1093/bioinformatics/bty633>\

Saitou, N., & Nei, M. (1987). The neighbor-joining method: A new method for reconstructing phylogenetic trees. *Molecular Biology and Evolution, 4*(4), 406–425. <https://doi.org/10.1093/oxfordjournals.molbev.a040454>.\

Toparslan, E., Karabag, K., & Bilge, U. (2020). A workflow with R: Phylogenetic analyses and visualizations using mitochondrial cytochrome b gene sequences. *PLOS ONE, 15*(12), e0243927. <https://doi.org/10.1371/journal.pone.0243927>\

Yu, G., Smith, D. K., Zhu, H., Guan, Y., & Lam, T. T. (2017). ggtree: An r package for visualization and annotation of phylogenetic trees with their covariates and other associated data. *Methods in Ecology and Evolution, 8*(1), 28–36. <https://doi.org/10.1111/2041-210X.12628>\
